### ISSUE TITLE
[OpInfo] Make it easier to add OpInfos

### ISSUE THREAD

**Issue by eellison** [pytorch/pytorch#57577]
## 🚀 Feature 

OpInfos are a very helpful resource to test the robustness of a new operator in the pytorch ecosystem. By adding opinfos, you are getting a huge amount of testing for free. However, it would be nice if it were easier to add a new OpInfo so that we can migrate more of the codebase over. In particular, it would be nice if more properties were opt-in, or allowed an opt-in mechanism, instead of requiring specification at declaration sight. 

For example, I am adding `torch.nn.functional.hardswish`. I was not involved in adding the operator to the library and it would be nice if I had an easy mechanism to use its current dtype coverage in the test. One possibility would be manually specifying `dtypes=None`, which would then omit the tests for the operator for all dtypes which are unsupported. Especially if you are developing on a cpu machine, it's not always clear what dtypes are supported, e.g. bfloat16 ?. For me it was a pretty tedious process to get the dtypes correctly specified. 

Another API which might make it easier to add opinfos is something similar to the `--accept` mechanism, which might run your operator tests and print out the right OpInfo registration that would pass successfully. I had to comb through and add `supports_out=False` and `supports_gradgrad=False`. 

Again, OpInfos tremendously useful, I think it is worth considering finding ways to make adding new tests less friction even at the potential cost of not having as much information specified. Some of this may just be the result of a first-time user of the API which is understandable. 

cc @mruberry

**Comment by mruberry** [comment]
Having a tool that suggests metadata is a cool idea, @eellison, and I think we could create one that would describe most (all?) of the current metadata, too. But if you're running on a machine without a CUDA device we won't be able to describe the metadata needed for CUDA, of course. 

A mechanism like "run on the dtypes you detect are supported" would be convenient, but we would lose being able to see which dtypes an operator supported, and if an op stopped supporting a dtype its tests would silently pass. This part might require more thinking. Maybe we could use an expect file pattern?

**Comment by eellison** [comment]
> but we would lose being able to see which dtypes an operator supported, and if an op stopped supporting a dtype its tests would silently pass

How common is this for unary operators ? I think it is reasonable to add a new op to OpInfos for an existing operator for all the other benefits you are getting besides dtype testing without requiring the additional benefit that dtype coverage doesnt regress, which should have been tested in the original PR anyway. The biggest downside atm here is that I can't reasonably suggest OSS users / bootcampers to add OpInfos if they dont have a gpu machine. 

The accept file mechanism would be cool, but yea, as you said doesn't help without CUDA.

**Comment by mruberry** [comment]
Good point.

For engineers without GPU access they could use the CI to probe the dtypes. For operations which aren't composites they can also lookup its CUDA dispatch macro and mirror it in the OpInfo.

That said, I can supporting unblocking people by supporting, like `dtypesIfCUDA="programmatic"` or a similar keyword that indicates the runtime should determine the dtypes an operation supports dynamically. This wouldn't be too tricky to support. test_supported_dtypes would need a revision:

https://github.com/pytorch/pytorch/blob/126ea1ccad0d46d77b5225dcef16ca9b40d7964b/test/test_ops.py#L54

maybe into a helper function that returned the supported dtypes and compared them against the operator dtypes directly? That helper function could then be used to populate the OpInfos dtypes, too. And if the dtypes were acquired dynamically or programmatically then test_supported_dtypes could be skipped, too.

**Comment by Krovatkin** [comment]
@mruberry I'd like to propose another option that is partially orthogonal, partially overlapping with what you guys discussed. 
I'm running into an issue where a test doesn't pass a backward for a specific datatype:

```
hardtanh_backward_cuda" not implemented for 'BFloat16
```

And of course disabling the datatype gives me another error: `RuntimeError isn't raised`.

Could we just add a central option in `OpInfo` to skip the exact test names and make every test in `test_ops.py` to test for it?

**Comment by mruberry** [comment]
> @mruberry I'd like to propose suggest option that is partially orthogonal, partially overlapping with what you guys discussed.
> I'm running into an issue where a test doesn't pass a backward for a specific datatype:
> 
> ```
> hardtanh_backward_cuda" not implemented for 'BFloat16
> ```
> 
> And of course disabling the datatype gives me another error: `RuntimeError isn't raised`.
> 
> Could we just add a central option in `OpInfo` to skip the exact test names and make every test in `test_ops.py` to test for it?

This is a tricky case. But we have two solutions:

- @walterddr added the ability to specify which datatypes an op supports computing a gradient for, too
- we do have skip mechanisms

Skips are defined using this class:

https://github.com/pytorch/pytorch/blob/6404184700159213f7d64df62537e238822f8b15/torch/testing/_internal/common_methods_invocations.py#L70

Example use of skips:

https://github.com/pytorch/pytorch/blob/6404184700159213f7d64df62537e238822f8b15/torch/testing/_internal/common_methods_invocations.py#L3691

I'm also working on an elaborated note on what OpInfos do and how to work with them here:

https://github.com/pytorch/pytorch/pull/57428

I'll be sure to incorporate any feedback/suggestions on what we could document or explain better on that PR, so please take a look!


Does that actually address the issue you're hitting, @Krovatkin?

**Comment by mruberry** [comment]
Update from offline follow-up with @kshitij12345:

- @kshitij12345 is going to start developing a tool that will automate OpInfo construction
- I created an RFC for letting OpInfos request their dtypes programmatically here: https://github.com/pytorch/pytorch/issues/58292

The OpInfos would probably acquire their dtypes programmatically from the tool. Please share your thoughts on the RFC, @eellison @Krovatkin.
